src/ast/ast_rewrite_star_projections.c

```diff
@@ -0,0 +1,332 @@
+		if(type == CYPHER_AST_WITH) {
+			// The WITH clause contains either aliases or its own STAR projection.
+			_collect_with_projections(clause, &aliases);
+		} else if(type == CYPHER_AST_MATCH) {
+			// The MATCH clause contains one pattern of N paths.
+			const cypher_astnode_t *pattern = cypher_ast_match_get_pattern(clause);
+			_collect_aliases_in_pattern(pattern, &aliases);
+		} else if(type == CYPHER_AST_CREATE) {
+			// The CREATE clause contains one pattern of N paths.
+			const cypher_astnode_t *pattern = cypher_ast_create_get_pattern(clause);
+			_collect_aliases_in_pattern(pattern, &aliases);
+		} else if(type == CYPHER_AST_MERGE) {
+			// The MERGE clause contains one path.
+			const cypher_astnode_t *path = cypher_ast_merge_get_pattern_path(clause);
+			_collect_aliases_in_path(path, &aliases);
+		} else if(type == CYPHER_AST_UNWIND) {
+			// The UNWIND clause introduces one alias.
+			const cypher_astnode_t *unwind_alias = cypher_ast_unwind_get_alias(clause);
+			array_append(aliases, unwind_alias);
+		} else if(type == CYPHER_AST_CALL) {
+			_collect_call_projections(clause, &aliases);
+		}
+	}
+
+	// Sort and unique the aliases array so that we won't make redundant projections given queries like:
+	// MATCH (a)-[]->(a) RETURN *
+	_uniqueArray(aliases);
+
+	return aliases;
+}
+
+static void replace_clause
+(
+	cypher_astnode_t *root,
+	cypher_astnode_t *clause,
+	int scope_start,
+	int idx
+) {
+	// collect all aliases defined in this scope
+	const cypher_astnode_t **aliases = _collect_aliases_in_scope(root,
+																 scope_start, idx);
+	uint alias_count = array_len(aliases);
+	cypher_astnode_type_t t = cypher_astnode_type(clause);
+	cypher_astnode_t *projections[alias_count];
+
+	// convert aliases to expressions
+	for(uint i = 0; i < alias_count; i++) {
+		cypher_astnode_t *expression = cypher_ast_clone(aliases[i]);
+		const cypher_astnode_t *alias = NULL;
+		cypher_astnode_t **children = &expression;
+		unsigned int nchildren = 1;
+		struct cypher_input_range range = cypher_astnode_range(aliases[i]);
+
+		projections[i] = cypher_ast_projection(expression, alias, children,
+											   nchildren, range);
+	}
+
+	if(alias_count == 0) {
+		if(t == CYPHER_AST_RETURN) {
+			// error if this is a RETURN clause with no aliases
+			ErrorCtx_SetError("RETURN * is not allowed when there are no variables in scope");
+			array_free(aliases);
+			return;
+		} else {
+			// build an empty projection
+			// to make variable-less WITH clauses work:
+			// MATCH () WITH * CREATE ()
+			struct cypher_input_range range = { 0 };
+			// build a null node to project and an empty identifier as its alias
+			cypher_astnode_t *expression = cypher_ast_null(range);
+			cypher_astnode_t *identifier = cypher_ast_identifier("",
+																 1, range);
+			cypher_astnode_t *children[2];
+			children[0] = expression;
+			children[1] = identifier;
+			projections[0] = cypher_ast_projection(expression, identifier,
```

Comment:
I think this is an invalid memory access: `cypher_astnode_t *projections[alias_count];`
Due to: `if(alias_count == 0)`and `cypher_astnode_t *projections[alias_count];`